packages/strapi-plugin-graphql/services/Loaders.js

```diff
@@ -0,0 +1,187 @@
+'use strict';
+
+/**
+ * Loaders.js service
+ *
+ * @description: A set of functions similar to controller's actions to avoid code duplication.
+ */
+
+const _ = require('lodash');
+const DataLoader = require('dataloader');
+
+module.exports = {
+  loaders: {},
+
+  initializeLoader: function() {
+    // Create loaders for each relational field (exclude core models).
+    Object.keys(strapi.models)
+      .filter(model => model !== 'core_store')
+      .forEach(model => {
+        (strapi.models[model].associations || []).forEach(association => this.createLoader(association.collection || association.model, association.plugin));
+      });
+
+    // Reproduce the same pattern for each plugin.
+    Object.keys(strapi.plugins).forEach(plugin => {
+      Object.keys(strapi.plugins[plugin].models).forEach(model => {
+        (strapi.plugins[plugin].models[model].associations || []).forEach(association => this.createLoader(association.collection || association.model, association.plugin));
+      });
+    });
+  },
+
+  createLoader: function(model, plugin) {
+    const name = plugin ? `${plugin}__${model}`: model;
+
+    this.loaders[name] = new DataLoader(keys => {
+      return new Promise(async (resolve, reject) => {
+        try {
+          // Extract queries from keys and merge similar queries.
+          const { queries, map } = this.extractQueries(model, _.cloneDeep(keys));
+          // Run queries in parallel.
+          const results = await Promise.all(queries.map((query) => this.makeQuery(model, query)));
+          // Use to match initial queries order.
+          const data = this.mapData(model, keys, map, results);
+
+          resolve(data);
+        } catch (e) {
+          reject(e);
+        }
+      });
+    }, {
+      cacheKeyFn: (key) => {
+        return _.isObjectLike(key) ? JSON.stringify(_.cloneDeep(key)) : key;
+      }
+    });
+  },
+
+  mapData: function(model, originalMap, map, results) {
+    // Use map to re-dispatch data correctly based on initial keys.
+    return originalMap.map((query, index) => {
+      // Find the index of where we should extract the results.
+      const indexResults = map.findIndex(queryMap => queryMap.indexOf(index) !== -1);
+      const data = results[indexResults];
+
+      // Retrieving referring model.
+      const ref = this.retrieveModel(model, query.options.source);
+
+      if (query.single) {
+        // Return object instead of array for one-to-many relationship.
+        return data.find(entry => entry[ref.primaryKey].toString() === (query.params[ref.primaryKey] || '').toString());
+      }
+
+      // Extracting ids from original request to map with query results.
+      const ids = (() => {
```

Comment:
I think that you don't need to create a IIFE, you can just do something like this:
```JS
// Single object to retrieve (one-to-many).
const alias = _.keys(query.options.query)[0];

// Extracting ids from original request to map with query results.
const ids = _.get(query.options, `query.${ref.primaryKey}`, {
    alias,
    value: _.get(query.options, `query.${alias}`)
});
```
And probably be more explicit about your variable naming

Reply:
The best option would be to create a new function to extract the IDs in that case. I will do that, thank you for pointing this out 👍